syntax = "proto3";

package sbai_behavior_protos;

import "sbai_geographic_protos/geo_point.proto";

enum Behavior {
    SURROUND = 0;
    CONVERGE = 1;
    DEFEND = 2;
    LINE_FORMATION = 3;
    RAPTOR = 4;
    NAVIGATE_TO_WAYPOINT = 5;
    AREA_COVERAGE = 6;
}

message SurroundBehaviorParameters {
    sbai_geographic_protos.GeoPoint surround_point = 1;
    float surround_radius_m = 2;
}

message ConvergeBehaviorParameters {
    sbai_geographic_protos.GeoPoint converge_point = 1;
    float converge_safety_radius_m = 2;
}

message RaptorBehaviorParameters {
    sbai_geographic_protos.GeoPoint raptor_point = 1;
    float outer_radius_m = 2;
    float inner_radius_m = 3;
}

message LineFormationParameters {
    sbai_geographic_protos.GeoPoint reference_point = 1;
    float separation_distance_m = 2;
    float line_yaw_deg = 3;  // 0 degrees is due East; counter-clockwise
    float robot_yaw_deg = 4;  // Angle of robots in relation to the line
}

message DefendBehaviorParameters {
    sbai_geographic_protos.GeoPoint defend_point = 1;
    float defend_radius_m = 2;
}

message NavigateToWaypointParameters {
    sbai_geographic_protos.GeoPoint desired_waypoint = 1;
    float desired_yaw_deg = 2;
}

message AreaCoverageBehaviorParameters {
    repeated sbai_geographic_protos.GeoPoint coverage_area = 1;
    float lane_width_m = 2;
}

message BehaviorRequest {
    // Number of agents to assign to this behavior
    uint32 number_of_assigned_agents = 1;
    Behavior requested_behavior = 2;

    // Behavior Specific Parameters
    oneof behavior_params {
        SurroundBehaviorParameters surround = 3;
        ConvergeBehaviorParameters converge = 4;
        DefendBehaviorParameters defend = 5;
        LineFormationParameters linear_formation = 6;
        RaptorBehaviorParameters raptor = 7;
        NavigateToWaypointParameters waypoint_navigation = 8;
        AreaCoverageBehaviorParameters area_coverage = 9;
        // Add other behavior-specific parameter messages here as needed
    }
}
